### PR TITLE
[DependencyInjection] fix performances in circular refs detection

### LINKED ISSUE #37850: Injecting certain services causes O(n) compilation complexity, while injecting Container is O(1)
**Symfony version(s) affected**: 4.4.11

**Description**  
Injecting services with complicated service graphs causes large compilation slowdown compared to simply injecting container. 

This comes as a surprise, since injecting services directly comes as universal recommendation, people are instructed to do it in migration guides from Symfony 3.x to 4.x and is even recommended for performance reasons, while I observe opposite.

**How to reproduce**  
1. Download [project.zip](https://github.com/symfony/symfony/files/5081372/project.zip)
1. `composer install`
1. `time bin/console`

Notice the time. In my case this is around **10s** without xdebug and **1m:5s** with xdebug.

Now do
```diff
Index: project/src/AppBundle/Controller/Admin/Settings/FormFieldTypeController.php
===================================================================
@@ -12,9 +12,9 @@
 class FormFieldTypeController extends AbstractController
 {
     public function __construct(
-        PanelHandlerAll $panelHandlerAll,
-        CustomPanelLoader $customPanelLoader
-//        ContainerInterface $container
+        ContainerInterface $container
     ) {
     }
 }

```

Run `time bin/console` again. Notice the time. In my case this is around **3s** without xdebug and around **15s** with xdebug.

Now, why did I mention this is O(n) vs O(1) problem:
In `Resources/config/services/_default.yml` file is `c1`..`c26` service defined. In case of injecting service container, it doesn't really matter how many of them there are, compilation speed is pretty much constant. However in other case, speed is very much dependant on number of these service definitions.

**Additional context**  
I originally raised this issue on Slack. @nicolas-grekas encouraged me to create this issue ;) I was working on converting the retrieval of our public services to injecting them directly, as officially recommended as Symfony best practice. In middle of this I discovered compilation times got much worse than originally, as opposed to just injecting the ContainerInterface as was done before. I have spent lot of time gutting our application so I can provide reproducer, I hope it's useful, even though I know it's not perfect.

Possibly related: #37540. However, marking the services private has no impact on performance in my case.

Blackfire link: https://blackfire.io/profiles/17769187-521c-48d3-8600-3de9adece809/graph (this is from before I gutted the application into reproducer)

### PR DESCRIPTION
**Pull request by jderusse**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #37850
| License       | MIT
| Doc PR        | -

This PR change the way circular references are detected.  And improve the project submitted in #37850 [by **86%**](https://blackfire.io/profiles/compare/cc4aa41d-5b63-4caa-9a1b-fc282dc5d64a/graph) (Build the container in 1.1 sec instead of 10)

Issue is: When a project contains a lot Circular Reference, the Dumper [spend a lot of time](https://blackfire.io/profiles/54f6eba2-b93c-4f2b-a268-1e58883faecb/graph) in merging those circular references.
note: a circular reference is not an issue when an service is not injected by constructor, but this Can not be known until all references are resolved (performed previously by connectCircularReferences)

This PR removed the connectCircularReferences and generate a flatten tree of dependencies:
 - the key is the service ID
 - the value is the list of direct **AND** indirect dependency + path to join the dependency 

I also [benched the PR with a project with few references](https://blackfire.io/profiles/compare/2f9902e6-3347-40b3-8421-e1fd09c067d2/graph) and result are almost the same before/after.

### CODE REVIEW

**Comment by ostrolucky** [comment]
I don't want to spam original issue regarding testing results of this PR. I've just retested this with the zip I attached in issue, I can confirm promising results: getting 4s with a patch and ~49s without this patch.


I've also retested my normal project from which I extracted the demo, there time didn't change much (which makes sense, I already replaced injections of these problematic services with container injections there). This time I used your [link to 4.4 file](https://github.com/jderusse/symfony/blob/perf-circular-44/src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php), perhaps during my previous testing I applied patch in this PR manually wrong and that resulted in some infinite loop.

**Comment by jderusse** [comment]
Thank you @ostrolucky for your tests.

This PR changes a lot of things in the way the container is Dumped, The output of the circularReferences before and after the PR are not comparable but the dumped container is identical in the cases I tested.

I think we should test this PR is several different configuration to be sure.

**Comment by jderusse** [comment]
Rework a little bit the implementation with the help of @nicolas-grekas 

Now reduced by 88% CPU time for the same memory usage https://blackfire.io/profiles/compare/ba4c464a-4449-4633-99a8-fcf31bcc948c/graph

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php [hunk 1]
```diff
@@ -409,58 +391,92 @@ private function getProxyDumper(): ProxyDumper
...
             }
         }
-        unset($currentPath[$sourceId], $subPath[$sourceId]);
+
+        $this->container->getCompiler()->getServiceReferenceGraph()->clear();
+        $this->singleUsePrivateIds = array_diff_key($this->singleUsePrivateIds, $this->circularReferences);
     }
 
-    private function addCircularReferences(string $id, array $currentPath, bool $byConstructor)
+    private function collectCircularReferences(string $sourceId, array $edges, array &$checkedNodes, array &$newNodes, array $path = [])
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: `: bool`

**Review by nicolas-grekas — APPROVED** [review]
That's really cool, thanks for having a deep look at this. Despite being labeled as a performance improvement, this really is a bug fix and should be for 4.4. The current logic follows too many paths and can take millions of iterations to complete.
The new logic doesn't inspect all possible loops, but it inspects the ones that matter.

:+1:

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php (hunk 1 above)
Comment (jderusse) [reply src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: fixed

**Comment by nicolas-grekas** [comment]
Thank you @jderusse.

**Comment by kocsismate** [comment]
Just a heads up, I've built the Symfony containers for https://github.com/kocsismate/php-di-container-benchmarks via PHP 8, running on Docker on my laptop:

- v5.1.8: `11805.0532 ms`
- 5.x-dev: `14068.2973 ms`

I ran the container building script multiple times, but the results stayed consistent (within +-1 sec).

**Comment by jderusse** [comment]
Hello @kocsismate , thanks for reporting these numbers.

You compared 2 differents version of the component (5.1 vs 5.2).
Could you please compare with version v5.2.0-BETA3

From my side I tried with PHP 7.4, and got ~similar numbers.
5.2.0-BETA3: `1928.255282 ms`
5.2-dev: `1995.516162 ms`

Note, I also extract number from the part that have been optimized (dumping the php code)
```diff
diff --git a/src/Container/Symfony/SymfonyContainerAdapter.php b/src/Container/Symfony/SymfonyContainerAdapter.php
index 4f7d174..69f4921 100644
--- a/src/Container/Symfony/SymfonyContainerAdapter.php
+++ b/src/Container/Symfony/SymfonyContainerAdapter.php
@@ -139,6 +139,8 @@ final class SymfonyContainerAdapter implements ContainerAdapterInterface
     protected function dumpContainer(ContainerBuilder $containerBuilder, string $path, string $class, bool $asFiles): void
     {
         $dumper = new PhpDumper($containerBuilder);
+
+        $s = microtime(true);
         $content = $dumper->dump(
             [
                 "namespace" => "DiContainerBenchmarks\\Container\\Symfony\\Resource",
@@ -147,6 +149,7 @@ final class SymfonyContainerAdapter implements ContainerAdapterInterface
                 "debug" => false,
             ]
         );
+        var_dump(microtime(true) - $s);
 
         if ($asFiles) {
             $file = key($content);
```

and get very similar number:
5.2.0-BETA3: `float(0.035419940948486) => 354 ms`
5.2-dev: `float(0.037868022918701) => 378 ms`

The result are ~ identical because your project does not have many circular references, thus you were not impacted by the bug.

**Comment by kocsismate** [comment]
Yeah, I can confirm, 5.2.0-BETA3 and 5.2-dev yields nearly the same results. Thanks for the explanation about the difference between my initial two measurements!

**Pull request merged**